odoo/api.py

```diff
@@ -753,14 +753,25 @@ def invalidate_all(self, flush=True):
 
     def _recompute_all(self):
         """ Process all pending computations. """
-        for field in list(self.fields_to_compute()):
-            self[field.model_name]._recompute_field(field)
+        # loop because a field's computation may trigger new computations
+        while True:
+            # fields to compute on real records (we can ignore new records)
+            fields_ = [field for field, ids in self.all.tocompute.items() if any(ids)]
+            if not fields_:
+                break
+            for field in fields_:
+                self[field.model_name]._recompute_field(field)
 
     def flush_all(self):
         """ Flush all pending computations and updates to the database. """
-        self._recompute_all()
-        for model_name in OrderedSet(field.model_name for field in self.cache.get_dirty_fields()):
-            self[model_name].flush_model()
+        # loop because flushing may trigger new computations and/or updates
+        while True:
+            self._recompute_all()
+            model_names = OrderedSet(field.model_name for field in self.cache.get_dirty_fields())
+            if not model_names:
+                break
+            for model_name in model_names:
+                self[model_name].flush_model()
```

Comment:
You can/should call directly `self._flush(fnames)` because `flush_model` recall `_recompute_model` a last time (but only for model that have at least one dirty field). I bet that behavior was already mimic a very bad 'fixpoint' implementation that depends on the dirty fields and the order of `model_names`.